src/hotspot/share/asm/register.hpp

```diff
@@ -245,14 +253,35 @@ inline ReverseRegSetIterator<RegImpl> AbstractRegSet<RegImpl>::rbegin() {
 
 // Debugging support
 
+template<typename R, typename... Rx>
+inline constexpr bool different_registers(AbstractRegSet<R> allocated_regs, R first_register) {
+  return !allocated_regs.contains(first_register);
+}
+
+template<typename R, typename... Rx>
+inline constexpr bool different_registers(AbstractRegSet<R> allocated_regs, R first_register, Rx... more_registers) {
+  if (allocated_regs.contains(first_register)) {
```

Comment:
FWIW, while first reading this I was looking for the base case of the recursion (the previous versions had some extra specializations). To me it looks like the base case is written in both this function and the function above. I would prefer to have the implementation inside one function only and change this function to use:
```
  if (!different_registers(allocated_regs, first_register)) {
```
I think this could make it a bit clearer, but if you prefer the current style, I think that's fine as well.

Reply:
I'd prefer to stick with what I have, because it's a bit more direct and slightly simpler runtime code.